### PR TITLE
afpd: eliminate per-entry rfork probes and per-op getcwd on hot paths

### PR DESCRIPTION
**Pull request by andylemin**
Two syscall eliminations on the metadata hot paths, each verified with A/B strace against an unmodified build on identical container workloads.

## Wiring up the negative side of the AD cache

The dircache's Tier 1 AD metadata cache and Tier 2 rfork data cache were built to serve *positive* answers: metadata fields and fork bytes we know exist. This PR wires up the negative signals — when we know there is *nothing* to cache, that confirmed absence is itself the cached answer, and serving it must cost zero filesystem contact.

`dcache_rlen` now carries the full state contract as named defines:

| state | value | metadata | rfork | what the cache answers |
|---|---|---|---|---|
| `AD_RLEN_NO_RFORK` | `0` | cached | absent | everything: Tier 1 fields, RFLEN 0 |
| fork length | `> 0` | cached | present | everything; Tier 2 eligible |
| `AD_RLEN_UNKNOWN` | `-1` | unknown | unknown | nothing: disk read on next access |
| `AD_RLEN_NO_AD` | `-2` | absent | absent | metadata ENOENT **and** RFLEN 0 |
| `AD_RLEN_RFORK_ONLY` | `-3` | absent | present | metadata ENOENT; fork size stays on-disk truth |

`AD_RLEN_NO_RFORK` is the floor of the positive range, not a sentinel — consumers test `>= 0` for "metadata cached". Only `NO_AD` authorizes the zero-contact RFLEN answer; `RFORK_ONLY` exists because with `ea = sys` the metadata and the resource fork are separate EAs, so files written by Samba's vfs_fruit or macOS-native tools carry a fork with no netatalk metadata — collapsing that state into either neighbour would each be wrong (`NO_AD` would pin their RFLEN at 0; `UNKNOWN` would re-read disk on every access).

**Minting.** The negative states are minted at the single point where the metadata read fails ENOENT: one `ad_reso_size()` probe classifies the miss — once **per ctime change**, never per read. `ad_reso_size()` returns 0 for *any* failure, so the probe's errno must corroborate absence before `NO_AD` is recorded; a transient error (EIO etc.) leaves `UNKNOWN` and retries next access rather than caching a false absence. Directories mint `NO_AD` unconditionally.

**Consuming.** All negative-state interpretation lives at one choke point (`ad_metadata_cached()`, where both `NO_AD` and `RFORK_ONLY` serve the metadata ENOENT from cache) plus the RFLEN/EXTRFLEN fast paths (which act only on `NO_AD`). `ad_getcomment()` previously bypassed the cache with a raw `ad_metadata()` call and now honours the same contract: a ctime/inode-validated entry with no metadata answers "no comment" without touching disk, skipped when an open fork may have just created metadata. Invalidation needs no new wiring — every existing reset path already writes `UNKNOWN`, which covers all sentinels.

**Guarding the positive side.** The opportunistic stores on the fork-owned adouble paths are gated on a new `ad_meta_loaded()` predicate: a fork opened on a file without metadata carries a zeroed header, and storing it would flip a negative sentinel into a positive entry serving zeroed FinderInfo (bypassing the extension-derived type/creator fallback). An fd-based open test is the wrong gate here — ea RDONLY metadata opens read the EA by path and hold no fd, so it would reject valid metadata on the common enumerate-with-open-fork path. Symlinks need explicit exclusion: both `ad_open_hf` backends short-circuit with the meta refcount already taken and no header read, so the predicate also rejects `AD_SYMLINK` on either fd (ea marks the meta fd, v2 the data fd). Alongside it, the `AD_DATA_OPEN`/`AD_META_OPEN`/`AD_RSRC_OPEN` predicate macros become static inline functions (`ad_data_open()`, `ad_meta_open()`, `ad_rsrc_open()`): type-checked, side-effect safe, and consistent with the new helper. The `ad_*_fileno` accessors stay macros — they are assigned to as lvalues.

**Coverage.** A unit test writes a resource fork, strips only the metadata EA (the foreign-writer state), and asserts both the reported RFLEN and the minted state — `RFORK_ONLY`, not `NO_AD`, not `UNKNOWN`.

## The rfork probe that survived every cache

`getmetadata()`'s RFLEN/EXTRFLEN branches fall back to `ad_reso_size()` — a sidecar stat (ea=ad) or rfork-EA probe (ea=sys) — whenever no adouble handle is present. But that no-handle case is routinely the negative-cached answer, and the correct reply is 0 with no filesystem contact.

This sits inside the per-entry reply loop of directory enumeration — the hottest workload a file server has. macOS Finder requests EXTRFLEN for every entry of every enumeration batch, nearly all files are fork-less, and the probe survives a fully warm cache: N syscalls per batch, per browse, forever, on all deployments. The fix consults the request's already-resolved dircache entry (threaded via `path->d_cached`, guarded by the `path_cached_file()` staleness accessor) and answers 0 when `NO_AD` is authoritative. `UNKNOWN`, `RFORK_ONLY`, and genuinely uncached files keep the on-disk probe; open-fork callers are untouched.

### Aligning lantest with Finder's request shape

The lantest Enumerate test did not reproduce this cost: it spoke AFP 2 — `FPEnumerate` with `LNAME` and 32-bit `DFLEN`/`RFLEN` — a request shape no modern client sends, which is precisely how this overhead stayed hidden from the dashboard while every real Finder browse paid it. The test now issues `FPEnumerateExt2` with Finder's bitmap (`PDINFO`, `EXTDFLEN`, `EXTRFLEN`, `UNIXPR`), the same request macOS sends per enumeration batch. The server registers that verb from AFP 3.1, so the supported `-3` (AFP 3.0) run keeps the legacy enumerate with the closest 3.0 bitmap. Batch size, interleave pattern, and op count are unchanged on both paths.

Two caveats on what the dashboard will show. The Enumerate metric's 9-PR history predates the verb change, so its first run breaks comparability — the median-shift baseline absorbs the common-mode part, but read that row's first delta as the test changing, not the code. And lantest creates its files through AFP, so every entry carries AD metadata and enumerate exercises the warm positive cache; the negative-cache win applies to foreign files (rsync'd trees, SMB-written shares) that a pure AFP client cannot create. The A/B evidence below therefore uses a Finder-shaped workload against fork-less files (afparg FPEnumerate with Finder's file bitmap, cache-hot, 500 fork-less files):

| cache-hot enumerate, 500 fork-less files | before | after | reduction |
|---|---|---|---|
| per-entry sidecar/AD probes | 1005 (~2/entry) | 500 (~1/entry) | **50%** |
| stat-family syscalls total | 2040 | 1535 | **25%** |

The remaining ~1 probe per entry comes from the FPGetFileDirParms path each entry also receives in this workload — a separate lookup shape in the FPGetFileDirParms pipeline, not the enumeration reply loop this PR targets.

## getcwd before the FCE/Spotlight gate

C evaluates arguments before the call, so every `fce_register(..., fullpathname(upath), ...)` site paid a getcwd() syscall per create/rename/delete even with FCE unconfigured and Spotlight off — the no-listener early exits live inside the callees. The parent directory's absolute path is already cached in the dircache entry: a shared `dir_event_path()` helper builds the event path from `d_fullpath` (getcwd-free) and falls back to `fullpathname()` when no entry is available. The FPCopyFile site already open-coded this pattern and is converted to the helper; the file-move FCE and Spotlight blocks, which each built the same source path independently, are combined into one.

Merging those file-move blocks also fixes their divergent NULL handling: on `bformat()` failure the FCE arm previously sent a MOVE event with a NULL source path while the Spotlight arm skipped — both now skip, never emitting a half-formed event.

`dir_event_path()` never emits a clipped event string: on snprintf truncation it falls back to `fullpathname()`, whose MAXPATHLEN bound matches the kernel's own path limit.

Sites deliberately not changed: `fullpathname()` inside `LOG(...)` argument lists — the LOG macro's level check gates evaluation lazily.

| create+delete of 2000 files | before | after |
|---|---|---|
| getcwd | 14013 | 6011 |

The remainder is `ochdir()`'s symlink-escape verification, which is load-bearing and stays.

Path equivalence is by construction — `v_path` is canonicalized with `realpath()` at volume creation, so `d_fullpath` is rooted at the same physical path getcwd reports — and proven empirically: an `fce_listen` capture of 54 move/delete events across the FPMoveAndRename and FPDelete spectest suites is byte-identical before and after, including the post-`deletecurdir` parent-path case.

## Testing

- Both spectest legs green (ea=sys 350/0, ea=ad 354/0)
- afpd unit suite green (76 ok), including the new fork-without-metadata state test
- FCE event-path correctness proven by listener capture equality (spectests don't assert FCE payloads)
- No wall-clock claims: these are syscall eliminations whose effect scales with syscall cost (containers, security modules, NFS-backed volumes)

### CODE REVIEW

File: etc/afpd/directory.c [hunk 1]
```diff
@@ -3318,6 +3318,18 @@ int renamedir(struct vol *vol,
...
 
+const char *dir_event_path(char *buf, size_t buflen,
+                           const struct dir *parent,
+                           const char *name)
+{
+    if (parent == NULL || parent->d_fullpath == NULL) {
+        return fullpathname(name);
+    }
+
+    snprintf(buf, buflen, "%s/%s", cfrombstr(parent->d_fullpath), name);
```
[the remark this answers is not in the document]
Comment (andylemin) [reply etc/afpd/directory.c]: The premise doesn't hold: v_path is canonicalized at volume creation — both creatvol() call sites receive realpath_safe(path) (netatalk_conf.c), so every d_fullpath is rooted at the physical path, the same thing getcwd() reports after chdir. A volume configured via a symlink gets its physical path in d_fullpath too. This is also verified empirically: an fce_listen capture of 54 move/delete events plus create events across the spectest suites is byte-identical before and after this change.

File: etc/afpd/file.c [hunk 2]
```diff
@@ -765,7 +765,16 @@ int getmetadata(const AFPObj *obj,
                     aint = htonl(adp->ad_rlen);
                 }
             } else {
-                rlen = ad_reso_size(path->u_name, 0, NULL);
+                /* The negative AD cache (dcache_rlen == -2, "confirmed no
+                 * AD exists") answers rlen = 0 without a filesystem probe;
+                 * -1 means unknown and keeps the on-disk fallback. */
+                const struct dir *cachedfile = path_cached_file(path);
+
+                if (cachedfile && cachedfile->dcache_rlen == (off_t) -2) {
```
[the remark this answers is not in the document]
Comment (andylemin) [reply etc/afpd/file.c]: Confirmed — good catch. On `ea = sys` the metadata EA and the resource-fork EA are separate stores, so a metadata ENOENT alone doesn't prove the file has no resource fork (files written by Samba's vfs_fruit or macOS-native tools carry a fork with no netatalk metadata, and the `-2` marker would have pinned their reported RFLEN at 0).

Fixed at the mint site rather than the read sites: `-2` is now only recorded when a one-time `ad_reso_size()` probe confirms the fork is absent too; a present fork leaves the cache at "unknown" so length reads keep their on-disk answer. The probe runs once per mint (i.e., once per ctime change), never per read, so the enumeration fast path is unchanged for the common no-AppleDouble case. Directories mint unconditionally, having no resource fork.

Covered by a new unit test that writes a resource fork, strips only the metadata EA, and asserts the reported RFLEN survives the negative-cache path.

**Comment by andylemin** [comment]
augment review

File: etc/afpd/ad_cache.c [hunk 3]
```diff
@@ -239,14 +239,28 @@ int ad_metadata_cached(const char *name, int flags, struct adouble *adp,
...
-        dir->dcache_rlen = (off_t) -2;
+        /* ENOENT above only proves the metadata is absent. With ea=sys the
+         * resource fork is a separate EA that can exist without metadata
+         * (Samba vfs_fruit, macOS-native writers), and -2 is consumed as
+         * "no AD at all" by the RFLEN fast paths — so it may only be
+         * minted when the rfork is absent too. One probe here, at the
+         * mint (once per ctime change); rfork present leaves -1 so
+         * length reads keep their on-disk answer. Directories have no
+         * resource fork. */
+        if ((flags & ADFLAGS_DIR) || ad_reso_size(name, 0, NULL) == 0) {
```
[the remark this answers is not in the document]
Comment (andylemin) [reply etc/afpd/ad_cache.c]: Confirmed. The mint now requires errno to corroborate absence: errno is cleared before the probe and AD_RLEN_NO_AD is only recorded when the zero result pairs with errno 0/ENOENT/ENOATTR. An ambiguous probe (EIO etc.) leaves the entry at AD_RLEN_UNKNOWN, so the next access retries on disk instead of serving a cached false absence. The negative states are also now named constants with the mixed metadata-absent/rfork-present case split out as AD_RLEN_RFORK_ONLY — see the state table in the PR description.

File: test/afpd/subtests.c [hunk 4]
```diff
@@ -376,3 +377,123 @@ int test005_getmetadata_open_fork_outlives_path(struct vol *vol)
...
+
+        return -1;
+    }
+
+    ad_init(&ad, vol);
+
+    if (ad_open(&ad, fpath,
+                ADFLAGS_DF | ADFLAGS_HF | ADFLAGS_RF | ADFLAGS_RDWR |
+                ADFLAGS_CREATE, 0600) != 0) {
+        return -1;
```
[the remark this answers is not in the document]
Comment (andylemin) [reply test/afpd/subtests.c]: Fixed — the ad_open() failure path now routes through the common cleanup label, which restores the saved cwd and closes the fd. Also tightened the other error paths in the same test while there: the fullpath bstring is NULL-checked and freed when dir_new() fails (it only takes ownership on success), and a dircache_add() failure now dir_free()s the entry instead of leaking it.

**Comment by andylemin** [comment]
augment review

**Comment by andylemin** [comment]
**Across all tests (of varying CI runners), the following metadata heavy tests are always faster by 10-20%**;
`Creating 2000 files`, `Create 2000 dirs tree (20×9×10)`, `Copying 1000 files client-side (R+W)`, `Copying 2000 files server-side`, `Deleting 2000 files`, `Mixed cache operations (create/stat/enum/delete) on 500 files`.

The `Enumerate dir with 2000 files` test has increased from ~5ms to ~14ms due to the change to use the modern `FPEnumerateExt2` (AFP Spec 3.1+) instead of the AFP2 only `FPEnumerate`. FPEnumerateExt2 before this change was 20ms+. So this change actually reduced FPEnumerateExt2 from 20 to 14ms.

Performance gains here were quite noticeable. 

*NB; The "Adj Δ%" column is proving to be useful now (getting better as more PRs land). For each push while working this PR, the CI runs overall performance varied from ~20% faster to ~20% slower (different CI hosts); but the "Adj Δ%" column reported surprisingly consistent/stable gains for the above mentioned tests which see improvements (within +-5% for every push). Metadata/CNID heavy tests (disk bound) still fluctuate heavily however, only the CPU bound ones are stable under Adj Δ% - this evidence is what has driven #3243 to detach CNID bound tests from CI disk performance*

File: include/atalk/adouble.h [hunk 5]
```diff
@@ -398,10 +398,28 @@ typedef enum {
...
+static inline int ad_rsrc_open(const struct adouble *ad)
+{
+    return ad->ad_reso_refcount && ad_reso_fileno(ad) >= 0;
+}
+
+/*! Metadata header read into ad; unlike ad_meta_open(), true also for
+ * ea RDONLY opens, which read the EA by path and hold no fd. */
+static inline int ad_meta_loaded(const struct adouble *ad)
+{
+    return ad->ad_meta_refcount > 0;
```
[the remark this answers is not in the document]
Comment (andylemin) [reply include/atalk/adouble.h]: Confirmed — both ad_open_hf backends short-circuit for symlinks after the meta refcount is taken and before any header read, so the refcount alone misreports a no-follow symlink as loaded. ad_meta_loaded() now also rejects AD_SYMLINK on either fd (ea marks the meta fd, v2 the data fd), so the opportunistic stores can no longer replace a cache entry with a symlink's zeroed header.

File: test/testsuite/lantest.c [hunk 6]
```diff
@@ -1446,45 +1446,40 @@ void run_test(const int32_t dir)
...
-        if (FPEnumerateFull(Conn, vol, 1, 40, DSI_DATASIZ, dir, "",
-                            (1 << FILPBIT_LNAME) | (1 << FILPBIT_FNUM) | (1 << FILPBIT_ATTR) |
-                            (1 << FILPBIT_FINFO) | (1 << FILPBIT_CDATE) | (1 << FILPBIT_BDATE) |
-                            (1 << FILPBIT_MDATE) | (1 << FILPBIT_DFLEN) | (1 << FILPBIT_RFLEN)
-                            ,
-                            (1 << DIRPBIT_ATTR) | (1 << DIRPBIT_ATTR) | (1 << DIRPBIT_FINFO) |
-                            (1 << DIRPBIT_CDATE) | (1 << DIRPBIT_BDATE) | (1 << DIRPBIT_MDATE) |
-                            (1 << DIRPBIT_LNAME) | (1 << DIRPBIT_PDID) | (1 << DIRPBIT_DID) |
-                            (1 << DIRPBIT_ACCESS))) {
+        if (FPEnumerateExt2Full(Conn, vol, dir, "",
```
[the remark this answers is not in the document]
Comment (andylemin) [reply test/testsuite/lantest.c]: Confirmed — afp_enumerate_ext2 is registered from the 31 case in the postauth switch, so a -3 (AFP 3.0) session gets AFPERR_NOOP and the benchmark aborted. The Enumerate test now gates on the session version: 3.1+ sends FPEnumerateExt2 with the Finder bitmap, and 3.0 keeps the legacy FPEnumerate with the closest 3.0 bitmap (DFLEN/RFLEN/LNAME in place of the Ext2-only bits). Batch pattern and op count are the same on both paths.

**Comment by andylemin** [comment]
@rdmark ready for review

**Review by rdmark — APPROVED** [review]
